pie/scripts/tune.py

```diff
@@ -0,0 +1,195 @@
+    infix = '+'.join(get_targets(settings)) + '-' + timestamp
+    return fname, infix
+
+
+def save(checkpoint_dir, settings, model):
+    # save model
+    fpath, infix = get_fname_infix(settings)
+    fpath = os.path.join(fpath, "checkpoint-"+str(checkpoint_dir))
+    os.makedirs(fpath, exist_ok=True)
+    fpath = model.save(fpath, infix=infix, settings=settings)
+    return fpath
+
+
+def env_setup(settings):
+    now = datetime.now()
+    # set seed
+    seed = now.hour * 10000 + now.minute * 100 + now.second
+    print("Using seed:", seed)
+    random.seed(seed)
+    numpy.random.seed(seed)
+    torch.manual_seed(seed)
+    if torch.cuda.is_available():
+        torch.cuda.manual_seed(seed)
+
+    if settings.verbose:
+        logging.basicConfig(level=logging.INFO)
+
+
+def affect_settings(target: Dict[str, Any], key: str, value):
+    if "/" in key:
+        i = key.index("/")
+        subkey, key = key[:i], key[i+1:]
+        # Would be cool to be able to filter with `[KEY=Value]` in a list, specifically for target=True
+        if subkey not in target:
+            target[subkey] = {}
+        target[subkey] = affect_settings(target[subkey], key, value)
+    else:
+        target[key] = value
+    return target
+
+
+def get_pruner(pruner_settings: Dict[str, Any]):
+    return getattr(optuna.pruners, pruner_settings["name"])(
+        *pruner_settings.get("args", []),
+        **pruner_settings.get("kwargs", {})
+    )
+
+
+# https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/async_hyperband_example.py
+
+def create_tuna_optimization(trial: optuna.Trial, fn: str, name: str, value: List[Any]):
+    """ Generate tuna value generator
+
+    Might use self one day, so...
+
+    :param trial:
+    :param fn:
+    :param name:
+    :param value:
+    :return:
+    """
+    return getattr(trial, fn)(name, *value)
+
+
+class Optimizer(object):
+    def __init__(self, settings, optimization_settings: List[Dict[str, Any]], gpus: List[int] = []):
+        self.focus: str = [task["name"] for task in settings.tasks if task.get("target") is True][0]
+        self.settings = settings
+        self.optimization_settings = optimization_settings
+        self.gpus = gpus
+
+    def optimize(self, trial: optuna.Trial):
+        env_setup(self.settings)
+
+        settings = self.settings
+
+        for opt_set in self.optimization_settings:
+            settings = affect_settings(
+                target=settings,
+                key=opt_set["path"],
+                value=create_tuna_optimization(
+                    trial=trial,
+                    fn=opt_set["type"],
+                    name=opt_set["path"].replace("/", "__"),
+                    value=opt_set["args"]
+                )
+            )
+
+        trainer, model, trainset, devset, label_encoder, reader = Trainer.setup(settings)
+
+        def report(epoch_id, _scores):
+            target = _scores[self.focus]["all"]["accuracy"]
```

Comment:
It'd be nice for the objective metric to be read off the optimization config file.

Reply:
You're right. Might be a use for the javascript path here...